lib/sqlalchemy/sql/functions.py

```diff
@@ -49,7 +56,38 @@ def register_function(identifier, fn, package="_default"):
 
     """
     reg = _registry[package]
-    reg[identifier] = fn
+    raw_identifier = identifier
+    identifier = identifier.lower()
+
+    # Check if a function with the same lowercase identifier is registered.
+    if identifier in _case_sensitive_reg[package]:
+        if (
+            raw_identifier not in _case_sensitive_reg[package][identifier]
+        ):
+            util.warn_deprecated(
+                "GenericFunction(s) '{}' are already registered with "
+                "different letter cases and might interact with {}.".format(
+                    _case_sensitive_reg[package][identifier],
+                    raw_identifier))
+            # If a function with the same lowercase identifier is registered,
+            # then these 2 functions are considered as case-sensitive.
+            if len(_case_sensitive_reg[package][identifier]) == 1:
```

Comment:
**mike bayer** (zzzeek) wrote:

the len() here makes me uncomfortable for two reasons.  One is, it's much better if we can do boolean checks, like, "if collection:", rather than "if len(collection) == 1".   The other is, race conditions / thread safety - this code is not threadsafe because the length can grow while we are still checking it for length one, or two threads can try to call "del" at the same time.

So when I originally suggested a set(), I see that doesn't work because this is considering the first element in the list to be the official "this was the case insensitive version".   I think there's less chance of race complexity if we just mirror all the case-sensitive names/ functions here, so that the transfer to "_registry" with case sensitive can be done idempotently.

if we store all the fns in both registries, then this logic should be able to be:

   case_sensitive_reg = _case_sensitive_registry[package]

   # ...

   if case_sensitive_reg[identifier]:
        # remove case insensitive name
        old_fn = reg.pop(identifier, None)

        # populate with all case sensitive names/ fns
        reg.update(case_sensitive_reg[identifier])

do I have that right?   I've revised my thinking here several times so I could be in a fog by now.

Reply:
I also don't really like this `len()` but I did not figure out how I could achieve the same result with another way, except by writing a specific class to handle the function registry or by using a third registry with booleans to tag the case-sensitve and case-incensitive identifiers. Because here, the `len(...) == 1` means 'One (and only one) function with the given `identifier` was already registered and now we want to register a second one with a different letter case, so we must switch to case-sensitive mode'. But after that this second function have been registered, all the next functions with the same `identifier` must be considered as case-sensitive. If we just look if `identifier` is in `case_sensitive_reg`, it does not mean there is only one function registered, so we miss something. For example, if we use your snippet to register `MyFunc`, then `myfunc`, then `MYFUNC`, then the last one will pop the second one, which is not the intended result.

Sorry if I am not clear to describe this.